extensions/mypy_extensions.py

```diff
@@ -7,16 +7,78 @@
 
 # NOTE: This module must support Python 2.7 in addition to Python 3.x
 
+import sys
+import types
+from typing import _type_check  # type: ignore
 
-def TypedDict(typename, fields):
-    """TypedDict creates a dictionary type that expects all of its
+
+def _check_fails(cls, other):
+    try:
+        if sys._getframe(1).f_globals['__name__'] not in ['abc', 'functools']:
+            raise TypeError('TypedDict does not support instance and class checks')
+    except (AttributeError, ValueError):
+        pass
+    return False
+
+def _dict_new(inst, cls, *args, **kwargs):
+    return dict(*args, **kwargs)
+
+def _typeddict_new(inst, cls, _typename, _fields=None, **kwargs):
+    if _fields is None:
+        _fields = kwargs
+    elif kwargs:
+        raise TypeError("TypedDict takes either a dict or "
+                        "keyword arguments, but not both")
+    return _TypedDictMeta(_typename, (), {'__annotations__': dict(_fields)})
+
+class _TypedDictMeta(type):
+    def __new__(cls, name, bases, ns):
+        tp_dict = super(_TypedDictMeta, cls).__new__(cls, name, (dict,), ns)
+        try:
+            tp_dict.__module__ = sys._getframe(2).f_globals.get('__name__', '__main__')
+        except (AttributeError, ValueError):
+            pass
+        anns = ns.get('__annotations__', {})
+        msg = "TypedDict('Name', {f0: t0, f1: t1, ...}); each t must be a type"
+        anns = {n: _type_check(tp, msg) for n, tp in anns.items()}
+        for base in bases:
+            anns.update(base.__dict__.get('__annotations__', {}))
+        tp_dict.__annotations__ = anns
+        if name == 'TypedDict':
+            tp_dict.__new__ = types.MethodType(_typeddict_new, tp_dict)
```

Comment:
AFAIK the signature of `MethodType` is undocumented. Why not just remove the unused `inst` arg from `_typeddict_new` and `_dict_new` and plunk the bare function into `tp_dict.__new__`? It seems to work.